src/operator/subgraph/dnnl/dnnl_transformer_qk_common.h

```diff
@@ -0,0 +1,232 @@
+  if (seed_node.node->op() == Op::Get("batch_dot")) {
+    *status = kStart;
+    matched_list->clear();
+    matched_list->push_back(&seed_node);
+    return true;
+  }
+  return false;
+}
+
+template <mode qk_mode>
+bool SelectInput(SelectStatusTransformerQK* status,
+                 std::vector<const BiDirectedNode*>* matched_list,
+                 const BiDirectedNode& n,
+                 const BiDirectedNode& input_node) {
+  if (*status == kFail || *status == kSuccess || input_node.node->is_variable())
+    return false;
+  const auto& raw_input_node = *input_node.node;
+  switch (*status) {
+    case kStart:
+      if (raw_input_node.op() == Op::Get("SwapAxis")) {
+        if (CheckSwapAxisConditionsQK(input_node)) {
+          *status = kFirstSwapAx;
+          matched_list->push_back(&input_node);
+        }
+        return true;
+      }
+      break;
+    case kFirstSwapAx:
+      if (raw_input_node.op() == Op::Get("SwapAxis")) {
+        if (CheckSwapAxisConditionsQK(input_node)) {
+          *status = kSecondSwapAx;
+          matched_list->push_back(&input_node);
+          return true;
+        }
+      }
+      break;
+    case kSecondSwapAx:
+      if (raw_input_node.op() == Op::Get("_npx_reshape")) {
+        // input to reshape must be first or second output from split
+        if (CheckReshapeConditionsQK(input_node, 0) || CheckReshapeConditionsQK(input_node, 1)) {
+          *status = kFirstReshape;
+          matched_list->push_back(&input_node);
+          return true;
+        }
+      }
+      break;
+    case kFirstReshape:
+      if (raw_input_node.op() == Op::Get("_npx_reshape")) {
+        if (CheckReshapeConditionsQK(input_node, 0) || CheckReshapeConditionsQK(input_node, 1)) {
+          if constexpr (qk_mode == mode::include_split) {
+            *status = kSecondReshape;
+          } else {
+            *status = kSuccess;
+          }
+          matched_list->push_back(&input_node);
+          return true;
+        }
+      }
+      break;
+    case kSecondReshape:
+      if (raw_input_node.op() == Op::Get("_split_v2") &&
+          CheckSplitConditions(*matched_list, input_node)) {
+        *status = kSuccess;
+        return true;
+      }
+      break;
+    default:
+      *status = kFail;
+      return false;
+  }
+  return false;
+}
+
+inline std::vector<BiDirectedNode*> Filter(const SelectStatusTransformerQK& status,
+                                           const std::vector<const BiDirectedNode*>& matched_list,
+                                           const std::vector<BiDirectedNode*>& candidates) {
+  if (status != kSuccess) {
+    return std::vector<BiDirectedNode*>(0);
+  } else {
+    std::vector<BiDirectedNode*> ret;
+    for (auto i : matched_list) {
+      auto non_const_i = const_cast<BiDirectedNode*>(i);
+      if (std::find(candidates.begin(), candidates.end(), non_const_i) != candidates.end()) {
+        ret.push_back(non_const_i);
+      }
+    }
+    return ret;
```

Comment:
Why not just return candidates?

Reply:
Impossible as it causes errors due to using candidates outside its class.